cupy/linalg/einsum.py

```diff
@@ -231,119 +369,207 @@ def einsum(*operands):
-        raise ValueError('einstein sum subscripts string contains a \'.\' that'
-                         'is not part of an ellipsis (\'...\')')
-
-    match = re.match('^([a-zA-Z@,]*)(->[a-zA-Z@]*)?$', subscripts)
-    if not match:
-        raise ValueError('einstein sum subscript string does not contain '
-                         'proper \'->\' output specified')
-
-    input_subscripts = match.group(1)
-    if match.group(2):
-        output_subscript = match.group(2)[2:]
-
-        if output_subscript.count('@') >= 2:
-            raise ValueError(
-                'Two or more \'...\' ellipsis can\'t be used for'
-                'output subscript')
-
-        # For compatibility with numpy.
-        # numpy.einsum arrows inputs like 'i->i...'.
-        # In this case, `...` does not affect the einsum results.
-        if '@' not in input_subscripts and '@' in subscripts:
-            output_subscript = output_subscript.replace('@', '')
-
-        irregular_chars = set(output_subscript) - set(input_subscripts)
-        if irregular_chars:
-            pickup = list(irregular_chars)[0]
-            raise ValueError('einstein sum subscripts string included output '
-                             'subscript \'{}\' which never appeared in an '
-                             'input'.format(pickup))
-
-        count_dict = collections.Counter(output_subscript)
-        for key in count_dict:
-            if count_dict[key] == 1:
-                continue
-            raise ValueError('einstein sum subscripts string includes output '
-                             'subscript \'{}\' multiple times'.format(key))
+    input_subscripts, output_subscript, operands = \
+        _parse_einsum_input(operands)
+    assert isinstance(input_subscripts, list)
+    assert isinstance(operands, list)
+
+    dtype = kwargs.pop('dtype', None)
+
+    # casting = kwargs.pop('casting', 'safe')
+    casting_kwargs = {}  # casting is not supported yet in astype
+
+    optimize = kwargs.pop('optimize', False)
+    # assert optimize is False, "optimize: sorry"
+    if optimize is True:
+        optimize = 'greedy'
+    if kwargs:
+        raise TypeError("Did not understand the following kwargs: %s"
+                        % list(kwargs.keys))
+
+    result_dtype = cupy.result_type(*operands) if dtype is None else dtype
+    operands = [
+        cupy.asanyarray(arr)
+        for arr in operands
+    ]
+
+    input_subscripts = [
+        _parse_ellipsis_subscript(sub, k, ndim=arr.ndim)
+        for k, (sub, arr) in enumerate(zip(input_subscripts, operands))
+    ]
+
+    # Get length of each unique dimension and ensure all dimensions are correct
+    dimension_dict = {}
+    for tnum, term in enumerate(input_subscripts):
+        sh = operands[tnum].shape
+        for cnum, char in enumerate(term):
+            dim = sh[cnum]
+            if char in dimension_dict.keys():
+                # For broadcasting cases we always want the largest dim size
```

Comment:
Don't allow broadcasting by default.

The following code fails only when `xp=numpy` and `opt=True`.

```
import numpy
import cupy

for xp in (cupy, numpy):
    for opt in (False, True):
        res = xp.einsum('i,i->i', xp.array([1, 2]), xp.array([1]), optimize=opt)
        print(res)
```

Reply:
I'll make the test require `numpy!=1.14.0`.